### ISSUE TITLE
False positive - About JS isSanitizerGuard of TaintTracking::Configuration

### ISSUE THREAD

**Issue by retanoj** [github/codeql#11488]
On semmle.javascript.dataflow.TaintTracking::TaintTracking::Configuration, `isSanitizerGuard` makes `if` condition act like a sanitizer. 

I added `isSanitizerGuard` code snippet to `SqlInjectionQuery.dll`, and used `javascript/ql/src/Security/CWE-089/SqlInjection.ql` to check below code. I found a situation which may be a false positive alert.

## This is NO. 1 test code, no alert

```javascript
const Router = require('koa-router')
const {Sequelize} = require("sequelize");

new Router().get("/hello", (ctx) => {
    const { startTime } = ctx.query;
    
    // first time to check ${startTime}
    if (validNanDateTime(startTime)) {
        throw new Error(`invalid startTime ${startTime}`);
    }

    const conditions = ['1'];

    // first time to use ${startTime} variable
    if (startTime) {
        conditions.push(`starttime > ${startTime}`)
    }

    new Sequelize().query(`SELECT * FROM t WHERE ${conditions.join(' and ')}`, null);
});

function validNanDateTime(date) {
    const d = new Date(date);
    return Number.isNaN(d.getTime());
}
```

## This is NO. 2 test code, one alert on `startTime` variable

```javascript
const Router = require('koa-router')
const {Sequelize} = require("sequelize");

new Router().get("/hello", (ctx) => {
    const { startTime } = ctx.query;

    // first time to check ${startTime}
    if (validNanDateTime(startTime)) {
        throw new Error(`invalid startTime ${startTime}`);
    }

    const conditions = ['1'];

    // first time to use ${startTime} variable
    if (startTime) {
        conditions.push(`starttime > ${startTime}`)
    }

    // second time to use ${startTime} variable
    if (startTime) {
        conditions.push(`starttime > ${startTime}`)
    }

    new Sequelize().query(`SELECT * FROM t WHERE ${conditions.join(' and ')}`, null);
});

function validNanDateTime(date) {
    const d = new Date(date);
    return Number.isNaN(d.getTime());
}
```

## This is NO. 3 test code, no alert

```javascript
const Router = require('koa-router')
const {Sequelize} = require("sequelize");

new Router().get("/hello", (ctx) => {
    const { startTime } = ctx.query;

    // first time to check ${startTime}
    if (validNanDateTime(startTime)) {
        throw new Error(`invalid startTime ${startTime}`);
    }

    const conditions = ['1'];

    // first time to use ${startTime} variable
    if (startTime) {
        conditions.push(`starttime > ${startTime}`)
    }

    // second time to check ${startTime}
    if (validNanDateTime(startTime)) {
        throw new Error(`invalid startTime ${startTime}`);
    }

    // second time to use ${startTime} variable
    if (startTime) {
        conditions.push(`starttime > ${startTime}`)
    }

    new Sequelize().query(`SELECT * FROM t WHERE ${conditions.join(' and ')}`, null);
});

function validNanDateTime(date) {
    const d = new Date(date);
    return Number.isNaN(d.getTime());
}
```

It looks like `isSanitizerGuard` effect on the nearest tainted variable and only once?
In my opinion, NO. 2 should have no alert. No matter how many times `startTime` variable has been used.
Is this a false positive?

<br>
<br>

## Below is my modified SqlInjectionQuery.qll code

```java
// modified javascript/ql/lib/semmle/javascript/security/dataflow/SqlInjectionQuery.dll

/**
 * Provides a taint tracking configuration for reasoning about string based
 * query injection vulnerabilities
 *
 * Note, for performance reasons: only import this file if
 * `SqlInjection::Configuration` is needed, otherwise
 * `SqlInjectionCustomizations` should be imported instead.
 */

import javascript
import SqlInjectionCustomizations::SqlInjection

class CustomAdHocWhitelistCheckSanitizer extends TaintTracking::AdHocWhitelistCheckSanitizer {
  override predicate sanitizes(boolean outcome, Expr e) {
    outcome = [true, false] and             // here include both if-then branch and else branch
    e = this.getArgument(0).asExpr()
  }
}

/**
 * A taint-tracking configuration for reasoning about string based query injection vulnerabilities.
 */
class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "SqlInjection" }

  override predicate isSource(DataFlow::Node source) { source instanceof Source }

  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }

  override predicate isSanitizer(DataFlow::Node node) {
    super.isSanitizer(node) or
    node instanceof Sanitizer
  }

  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
    exists(LdapJS::TaintPreservingLdapFilterStep filter |
      pred = filter.getInput() and
      succ = filter.getOutput()
    )
  }

  // enable CustomAdHocWhitelistCheckSanitizer
  override predicate isSanitizerGuard(TaintTracking::SanitizerGuardNode guard) {
    guard instanceof CustomAdHocWhitelistCheckSanitizer
  }
}

```

<br>
<br>

CodeQL command-line toolchain release 2.11.2.
CodeQL repo branch codeql-cli/v2.11.2

**Comment by asgerf** [comment]
Hi @retanoj 👋 

Thanks for your bug report. Indeed the extra use of `startTime` in the middle here should not affect taint flow to the sink.

I believe I've identified and fixed the underlying issue in https://github.com/github/codeql/pull/11565, it just needs more testing before we can merge it.

----

Independent of the above issue, about the that sanitizer you used in the example:
```js
function validDateTime(date) {
    const d = new Date(date);
    return Number.isNaN(d.getTime());
}
```
- It seems to do the opposite what the name suggests, it returns true if the date is _invalid_, not valid.
- It is not sufficient for sanitisation. For example, this string could still cause SQL injection:
```js
'(1); DROP TABLE students; 1'
```

**Comment by retanoj** [comment]
> Hi @retanoj 👋
> 
> Thanks for your bug report. Indeed the extra use of `startTime` in the middle here should not affect taint flow to the sink.
> 
> I believe I've identified and fixed the underlying issue in #11565, it just needs more testing before we can merge it.
> 
> Independent of the above issue, about the that sanitizer you used in the example:
> 
> ```js
> function validDateTime(date) {
>     const d = new Date(date);
>     return Number.isNaN(d.getTime());
> }
> ```
> 
> * It seems to do the opposite what the name suggests, it returns true if the date is _invalid_, not valid.
> * It is not sufficient for sanitisation. For example, this string could still cause SQL injection:
> 
> ```js
> '(1); DROP TABLE students; 1'
> ```

Hi @asgerf 

Thank you for your replying.

Yes, `validDateTime` function name is confused. Maybe it should be called `invalidDateTime`, but `AdHocWhitelistCheckSanitizer` do not capture `invalid.*`. So now I change it to `validNanDateTime` 😄 .

And yes again, this sanitizer is not sufficient for sanitization. I found another payload like this:

`(1) union select (1), (2), (3);-- 1`

**Comment by retanoj** [comment]
hi @asgerf 

May I ask how to debug codeql executing process? Is there any way to make a breakpoint on codeql source code ?
